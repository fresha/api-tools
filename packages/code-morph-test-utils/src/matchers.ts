import {
  MatcherHintOptions,
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
  stringify,
} from 'jest-matcher-utils';
import * as prettier from 'prettier';

import { poorMansElixirFormat } from './elixir';

const EXPECTED_LABEL = 'Expected';
const RECEIVED_LABEL = 'Received';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-unused-vars
  namespace jest {
    interface Matchers<R> {
      toHaveFormattedElixirText(test: string): R;
      toHaveFormattedTypeScriptText(text: string): R;
    }
  }
}

type SourceFile = {
  getText(): string;
};

const compareText = (
  received: string,
  expected: string,
  matcherName: string,
): jest.CustomMatcherResult => {
  const matcherOptions: MatcherHintOptions = {
    comment: 'deep equality',
  };
  const pass = received === expected;

  const message = pass
    ? () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        `Expected: not ${printExpected(expected)}\n` +
        (stringify(expected) !== stringify(received)
          ? `Received:     ${printReceived(received)}`
          : '')
    : () =>
        // eslint-disable-next-line prefer-template
        matcherHint(matcherName, undefined, undefined, matcherOptions) +
        '\n\n' +
        printDiffOrStringify(expected, received, EXPECTED_LABEL, RECEIVED_LABEL, true);

  return { message, pass };
};

export const createMatchers = () => ({
  toHaveFormattedElixirText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    return compareText(
      poorMansElixirFormat(sourceFile.getText()),
      poorMansElixirFormat(expectedText),
      'toHaveFormattedElixirText',
    );
  },

  toHaveFormattedTypeScriptText: (
    sourceFile: SourceFile,
    expectedText: string,
  ): jest.CustomMatcherResult => {
    const options: prettier.Options = {
      parser: 'typescript',
      singleQuote: true,
      trailingComma: 'all',
    };
    const received = prettier.format(sourceFile.getText(), options);
    const expected = prettier.format(expectedText, options);

    return compareText(received, expected, 'toHaveFormattedTypeScriptText');
  },
});
