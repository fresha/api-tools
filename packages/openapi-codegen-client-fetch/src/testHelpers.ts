// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';

import type { ActionContext, Context } from './context';
import type { OpenAPIModel, OperationModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel): Context => {
  const base = createTSProjectTestContext(openapi);

  return {
    ...base,
    includeDeprecated: false,
    includedTags: new Set<string>(),
    excludedTags: new Set<string>(),
    transformRequest: 'kebab',
    transformResponse: 'camel',
  };
};

export const createActionTestContext = (
  operation: OperationModel,
  fileName: string,
): ActionContext => {
  const base = createTestContext(operation.root);
  const sourceFile = base.project.createSourceFile(fileName, '');

  return {
    ...base,
    sourceFile,
    operation,
  };
};
