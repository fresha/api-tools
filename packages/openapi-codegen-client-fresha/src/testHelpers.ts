// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { getAPIName, getRootUrlOrThrow } from '@fresha/openapi-codegen-utils';

import type { Context } from './context';
import type { OpenAPIModel } from '@fresha/openapi-model/build/3.0.3';

export const createTestContext = (openapi: OpenAPIModel, rootDir = '/'): Context => {
  const base = createTSProjectTestContext(openapi, rootDir);

  let apiRootUrl: string;
  try {
    apiRootUrl = getRootUrlOrThrow(openapi);
  } catch (err) {
    apiRootUrl = 'API_URL';
  }

  return {
    ...base,
    apiName: getAPIName(openapi),
    apiRootUrl,
  };
};
