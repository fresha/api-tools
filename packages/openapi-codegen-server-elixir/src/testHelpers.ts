import { Project } from '@fresha/code-morph-ex';
import { createRegistry } from '@fresha/json-api-model';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTestContext as createBaseTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory } from '@fresha/openapi-model/build/3.0.3';

import { Generator } from './parts';

import type { Context } from './context';

export const createTestContext = (phoenixApp: string, rootDir = '/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createBaseTestContext(openapi, rootDir);

  const project = new Project({
    rootDir,
    phoenixApp,
    useInMemoryFileSystem: true,
  });

  return {
    ...base,
    testObjectFactoryModuleName: `${project.getAppModuleName()}.Factory`,
    openapi,
    project,
    registry: createRegistry(),
  };
};

export const createGenerator = (phoenixApp = 'api_tools_web'): Generator => {
  const context = createTestContext(phoenixApp, '/');
  const generator = new Generator(context);
  return generator;
};
