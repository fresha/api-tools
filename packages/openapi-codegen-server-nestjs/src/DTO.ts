import assert from 'assert';
import path from 'path';

import { addClassValidatorImports, addDecorator } from './utils';

import type { Generator } from './Generator';
import type { Nullable } from '@fresha/api-tools-core';
import type { SchemaModel } from '@fresha/openapi-model/build/3.0.3';
import type { SourceFile } from 'ts-morph';

export class DTO {
  readonly generator: Generator;
  readonly className: string;
  readonly outputPath: string;
  private readonly schema: Nullable<SchemaModel>;
  private readonly tsSourceFile: SourceFile;

  constructor(generator: Generator, name: string, schema: Nullable<SchemaModel>) {
    this.generator = generator;
    this.className = name;
    this.outputPath = path.join(this.generator.outputPath, 'dto', `${this.className}.dto.ts`);
    this.schema = schema;
    this.tsSourceFile = this.generator.tsProject.createSourceFile(this.outputPath, '', {
      overwrite: true,
    });
  }

  generateCode(): void {
    const classDecl = this.tsSourceFile.addClass({
      name: this.className,
      isExported: true,
    });

    if (this.schema) {
      assert(this.schema.type === 'object');

      for (const [propName, propSchema] of this.schema.properties) {
        switch (propSchema.type) {
          case null: {
            const propDef = classDecl.addProperty({
              name: propName,
              type: 'unknown',
              hasQuestionToken: !this.schema.required.has(propName),
            });
            if (propSchema.allOf?.length || propSchema.oneOf?.length || propSchema.anyOf?.length) {
              addClassValidatorImports(this.tsSourceFile, 'ValidateNested');
              addDecorator(propDef, 'ValidateNested', undefined);
            }
            break;
          }
          case 'boolean': {
            const propDef = classDecl.addProperty({
              name: propName,
              type: 'boolean',
              hasQuestionToken: !this.schema.required.has(propName),
            });
            addClassValidatorImports(this.tsSourceFile, 'IsBoolean');
            addDecorator(propDef, 'IsBoolean', undefined);
            break;
          }
          case 'number': {
            const propDef = classDecl.addProperty({
              name: propName,
              type: propSchema.nullable ? 'number | null' : 'number',
              hasQuestionToken: !this.schema.required.has(propName),
            });
            if (propSchema.minimum != null) {
              addClassValidatorImports(this.tsSourceFile, 'Min');
              addDecorator(propDef, 'Min', propSchema.minimum);
            }
            if (propSchema.exclusiveMinimum != null) {
              addClassValidatorImports(this.tsSourceFile, 'Min');
              addDecorator(propDef, 'Min', propSchema.exclusiveMinimum);
            }
            if (propSchema.maximum != null) {
              addClassValidatorImports(this.tsSourceFile, 'Max');
              addDecorator(propDef, 'Max', propSchema.maximum);
            }
            if (propSchema.exclusiveMaximum != null) {
              addClassValidatorImports(this.tsSourceFile, 'Max');
              addDecorator(propDef, 'Max', propSchema.exclusiveMaximum);
            }
            addClassValidatorImports(this.tsSourceFile, 'IsInt');
            addDecorator(propDef, 'IsInt', undefined);
            break;
          }
          case 'string': {
            const propDef = classDecl.addProperty({
              name: propName,
              type: propSchema.nullable ? 'string | null' : 'string',
              hasQuestionToken: !this.schema.required.has(propName),
            });
            if (propSchema.minLength) {
              addClassValidatorImports(this.tsSourceFile, 'MinLength');
              addDecorator(propDef, 'MinLength', propSchema.minLength);
            }
            if (propSchema.maxLength) {
              addClassValidatorImports(this.tsSourceFile, 'MaxLength');
              addDecorator(propDef, 'MaxLength', propSchema.maxLength);
            }
            addClassValidatorImports(this.tsSourceFile, 'IsString');
            addDecorator(propDef, 'IsString', undefined);
            break;
          }
          case 'object': {
            const propDef = classDecl.addProperty({
              name: propName,
              type: 'object',
              hasQuestionToken: !this.schema.required.has(propName),
            });
            addDecorator(propDef, 'ValidateNested', undefined);
            break;
          }
          case 'array': {
            let itemTypeString = 'unknown[]';
            switch (propSchema.items?.type) {
              case 'boolean':
                itemTypeString = 'boolean[]';
                break;
              case 'integer':
              case 'number':
                itemTypeString = 'number[]';
                break;
              case 'string':
                itemTypeString = 'string[]';
                break;
              case 'object':
                itemTypeString = 'object[]';
                break;
              case undefined:
              case null:
                break;
              default:
                assert.fail(`Unsupported schema type ${String(propSchema.items?.type)}`);
            }

            const propDef = classDecl.addProperty({
              name: propName,
              type: itemTypeString,
              hasQuestionToken: !this.schema.required.has(propName),
            });
            if (propSchema.minItems != null) {
              addClassValidatorImports(this.tsSourceFile, 'ArrayMinSize');
              addDecorator(propDef, 'ArrayMinSize', propSchema.minItems);
            }
            if (propSchema.maxItems != null) {
              addClassValidatorImports(this.tsSourceFile, 'ArrayMaxSize');
              addDecorator(propDef, 'ArrayMaxSize', propSchema.maxItems);
            }
            addClassValidatorImports(this.tsSourceFile, 'IsArray');
            addDecorator(propDef, 'IsArray', undefined);
            break;
          }
          default:
            assert.fail(`Unexpected schema type: ${String(propSchema.type)}`);
        }
      }
    }
  }
}
