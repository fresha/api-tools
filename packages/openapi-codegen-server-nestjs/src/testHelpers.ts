import { Nullable } from '@fresha/api-tools-core';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createTSProjectTestContext } from '@fresha/openapi-codegen-test-utils';
import { OpenAPIFactory, SchemaModel } from '@fresha/openapi-model/build/3.0.3';

import { Context } from './context';
import { Generator } from './Generator';

export const createTestContext = (nestApp = 'web', outputPath = '/tmp/'): Context => {
  const openapi = OpenAPIFactory.create();
  const base = createTSProjectTestContext(openapi, outputPath);

  base.project.createSourceFile(
    `${outputPath}/${nestApp}.module.ts`,
    `import { Module } from '@nestjs/common';
    import { AppController } from './app.controller';
    import { AppService } from './app.service';

    @Module({
      imports: [],
      controllers: [AppController],
      providers: [AppService],
    })
    export class AppModule {}`,
  );

  return {
    ...base,
    nestApp,
    addDTO: jest.fn(),
  };
};

export const createGenerator = (nestApp = 'app', rootDir = '/'): Generator => {
  const context = createTestContext(nestApp, rootDir);

  let generator: Generator;
  // eslint-disable-next-line prefer-const
  generator = new Generator({
    ...context,
    addDTO: (name: string, schema: Nullable<SchemaModel>) => generator.addDTO(name, schema),
  });

  return generator;
};
