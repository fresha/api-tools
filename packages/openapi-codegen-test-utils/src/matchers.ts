import {
  MatcherHintOptions,
  matcherHint,
  printDiffOrStringify,
  printExpected,
  printReceived,
  stringify,
} from 'jest-matcher-utils';
import * as prettier from 'prettier';

import type { SourceFile } from 'ts-morph';

const EXPECTED_LABEL = 'Expected';
const RECEIVED_LABEL = 'Received';

// eslint-disable-next-line @typescript-eslint/no-namespace, @typescript-eslint/no-unused-vars
declare global {
  namespace jest {
    interface Matchers<R> {
      toHaveFormattedText(text: string): R;
    }
  }
}

export const createMatchers = () => {
  expect.extend({
    toHaveFormattedText: (sourceFile: SourceFile, expectedText: string): jest.CustomMatcherResult => {
      const options: prettier.Options = {
        parser: 'typescript',
        singleQuote: true,
        trailingComma: 'all',
      };
      const received = prettier.format(sourceFile.getText(), options);
      const expected = prettier.format(expectedText, options);

      const matcherName = 'toHaveFormattedText';
      const matcherOptions: MatcherHintOptions = {
        comment: 'deep equality',
      };
      const pass = received === expected;

      const message = pass
        ? () =>
            // eslint-disable-next-line prefer-template
            matcherHint(matcherName, undefined, undefined, matcherOptions) +
            '\n\n' +
            `Expected: not ${printExpected(expected)}\n` +
            (stringify(expected) !== stringify(received)
              ? `Received:     ${printReceived(received)}`
              : '')
        : () =>
            // eslint-disable-next-line prefer-template
            matcherHint(matcherName, undefined, undefined, matcherOptions) +
            '\n\n' +
            printDiffOrStringify(expected, received, EXPECTED_LABEL, RECEIVED_LABEL, true);

      return { message, pass };
    },
  });
};
