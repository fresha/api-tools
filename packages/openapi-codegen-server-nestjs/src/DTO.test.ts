import '@fresha/openapi-codegen-test-utils/build/matchers';

import { DTO } from './DTO';
import { createGenerator } from './testHelpers';

import type { Context } from './context';

test('construction', () => {
  const { context } = createGenerator('app', '/tmp');

  const dto = new DTO(context, 'SomeResponse', null);
  expect(dto.className).toBe('SomeResponse');
  expect(dto.outputPath).toBe('/tmp/dto/SomeResponse.dto.ts');
});

describe('serialization', () => {
  let context: Context = {} as Context;

  beforeEach(() => {
    context = createGenerator('app', '/var').context;
  });

  test('proper name and ouput path', () => {
    new DTO(context, 'SomeResponse2', null).generateCode();

    expect(context.project.getSourceFiles()).toHaveLength(2);
    expect(context.project.getSourceFileOrThrow('/var/app.module.ts')).toBeTruthy();
    expect(
      context.project.getSourceFileOrThrow('/var/dto/SomeResponse2.dto.ts'),
    ).toHaveFormattedTypeScriptText(`export class SomeResponse2 {}`);
  });

  test('primitive schemas + optionality', () => {
    const schema = context.openapi.components.setSchema('Error', 'object');
    schema.setProperties({
      optionalBool: 'boolean',
      requiredBool: { type: 'boolean', required: true },
      numeric: 'number',
      requiredNum: { type: 'number', required: true },
      textual: 'string',
      requiredText: { type: 'string', required: true },
    });

    new DTO(context, 'Response', schema).generateCode();

    expect(
      context.project.getSourceFileOrThrow('/var/dto/Response.dto.ts'),
    ).toHaveFormattedTypeScriptText(
      `import { Expose } from 'class-transformer';
      import { IsBoolean, IsInt, IsString } from 'class-validator';

      export class Response {
        @Expose()
        @IsBoolean()
        optionalBool?: boolean;

        @Expose()
        @IsBoolean()
        requiredBool: boolean;

        @Expose()
        @IsInt()
        numeric?: number;

        @Expose()
        @IsInt()
        requiredNum: number;

        @Expose()
        @IsString()
        textual?: string;

        @Expose()
        @IsString()
        requiredText: string;
      }`,
    );
  });

  test('numeric limits', () => {
    const schema = context.openapi.components.setSchema('Error', 'object');
    schema.setProperties({
      min: { type: 'number', minimum: 10 },
      minExclusive: { type: 'number', exclusiveMinimum: 15 },
      max: { type: 'number', maximum: 20 },
      maxExclusive: { type: 'number', exclusiveMaximum: 25 },
    });

    new DTO(context, 'Response', schema).generateCode();

    expect(
      context.project.getSourceFileOrThrow('/var/dto/Response.dto.ts'),
    ).toHaveFormattedTypeScriptText(
      `import { Expose } from 'class-transformer';
      import { Min, Max, IsInt } from 'class-validator';

      export class Response {
        @Expose()
        @Min(10)
        @IsInt()
        min?: number;

        @Expose()
        @Min(15)
        @IsInt()
        minExclusive?: number;

        @Expose()
        @Max(20)
        @IsInt()
        max?: number;

        @Expose()
        @Max(25)
        @IsInt()
        maxExclusive?: number;
      }`,
    );
  });

  test('string limits', () => {
    const schema = context.openapi.components.setSchema('Error', 'object');
    schema.setProperties({
      minLen: { type: 'string', minLength: 1 },
      maxLen: { type: 'string', maxLength: 10 },
    });

    new DTO(context, 'Response2', schema).generateCode();

    expect(
      context.project.getSourceFileOrThrow('/var/dto/Response2.dto.ts'),
    ).toHaveFormattedTypeScriptText(
      `import { Expose } from 'class-transformer';
      import { MinLength, MaxLength, IsString } from 'class-validator';

      export class Response2 {
        @Expose()
        @MinLength(1)
        @IsString()
        minLen?: string;

        @Expose()
        @MaxLength(10)
        @IsString()
        maxLen?: string;
      }`,
    );
  });

  test('object', () => {
    const resourceSchema = context.openapi.components.setSchema('Employee', 'object');
    resourceSchema.setProperties({
      type: { type: 'string', required: true, enum: ['1.0'] },
      id: { type: 'string', required: true },
      attributes: { type: 'object', required: true },
      relationships: { type: 'object', required: true },
    });

    resourceSchema.getPropertyOrThrow('attributes').setProperties({
      name: { type: 'string', required: true },
      age: 'number',
      active: 'boolean',
    });

    const relationshipsSchema = resourceSchema.getPropertyOrThrow('relationships');
    const providerRelationSchema = relationshipsSchema.setProperty('provider', {
      type: 'object',
      required: true,
    });
    const providerRelDataSchema = providerRelationSchema.setProperty('data', {
      type: 'object',
      required: true,
    });
    providerRelDataSchema.setProperties({
      type: { type: 'string', required: true },
      id: { type: 'string', required: true },
    });

    new DTO(context, 'Employee', resourceSchema).generateCode();

    expect(context.project.getSourceFileOrThrow('/var/dto/Employee.dto.ts'))
      .toHaveFormattedTypeScriptText(`
      import { Expose, Type } from 'class-transformer';
      import { IsInt, IsBoolean, IsString } from 'class-validator';

      export class Employee {
        @Expose()
        @IsString()
        type: string;

        @Expose()
        @IsString()
        id: string;

        @Expose()
        @Type(() => EmployeeAttributes)
        @ValidateNested()
        attributes: EmployeeAttributes;

        @Expose()
        @Type(() => EmployeeRelationships)
        @ValidateNested()
        relationships: EmployeeRelationships;
      }

      export class EmployeeAttributes {
        @Expose()
        @IsString()
        name: string;

        @Expose()
        @IsInt()
        age?: number;

        @Expose()
        @IsBoolean()
        active?: boolean;
      }

      export class EmployeeRelationships {
        @Expose()
        @Type(() => EmployeeRelationshipsProvider)
        @ValidateNested()
        provider: EmployeeRelationshipsProvider;
      }

      export class EmployeeRelationshipsProvider {
        @Expose()
        @Type(() => EmployeeRelationshipsProviderData)
        @ValidateNested()
        data: EmployeeRelationshipsProviderData;
      }

      export class EmployeeRelationshipsProviderData {
        @Expose()
        @IsString()
        type: string;

        @Expose()
        @IsString()
        id: string;
      }
    `);
  });

  test('array', () => {
    const schema = context.openapi.components.setSchema('Error', 'object');

    schema.setProperty('intArray', 'array').setItems('integer');

    new DTO(context, 'AnotherResponse', schema).generateCode();

    expect(
      context.project.getSourceFileOrThrow('/var/dto/AnotherResponse.dto.ts'),
    ).toHaveFormattedTypeScriptText(
      `import { Expose } from 'class-transformer';
      import { IsArray } from 'class-validator';

      export class AnotherResponse {
        @Expose()
        @IsArray()
        intArray?: number[];
      }`,
    );
  });
});
