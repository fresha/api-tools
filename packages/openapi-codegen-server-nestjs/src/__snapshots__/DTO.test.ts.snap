// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialization array 1`] = `
"import { Expose } from "class-transformer";
import { IsArray } from "class-validator";

export class AnotherResponse {
    
@Expose()
    @IsArray()
    intArray?: number[];
}
"
`;

exports[`serialization numeric limits 1`] = `
"import { Expose } from "class-transformer";
import { Min, Max, IsInt } from "class-validator";

export class Response {
    
@Expose()
    @Min(10)
    @IsInt()
    min?: number;
    
@Expose()
    @Min(15)
    @IsInt()
    minExclusive?: number;
    
@Expose()
    @Max(20)
    @IsInt()
    max?: number;
    
@Expose()
    @Max(25)
    @IsInt()
    maxExclusive?: number;
    
@Expose()
    @Min(10)
    @IsInt()
    intMin?: number;
    
@Expose()
    @Min(16)
    @IsInt()
    intMinExclusive?: number;
    
@Expose()
    @Max(20)
    @IsInt()
    intMax?: number;
    
@Expose()
    @Max(24)
    @IsInt()
    intMaxExclusive?: number;
}
"
`;

exports[`serialization object 1`] = `
"import { Expose, Type } from "class-transformer";
import { IsInt, IsBoolean, IsString } from "class-validator";

export class Employee {
    
@Expose()
    @IsString()
    type: string;
    
@Expose()
    @IsString()
    id: string;
    
@Expose()
    @Type(() => EmployeeAttributes)
    @ValidateNested()
    attributes: EmployeeAttributes;
    
@Expose()
    @Type(() => EmployeeRelationships)
    @ValidateNested()
    relationships: EmployeeRelationships;
}

export class EmployeeAttributes {
    
@Expose()
    @IsString()
    name: string;
    
@Expose()
    @IsInt()
    age?: number;
    
@Expose()
    @IsBoolean()
    active?: boolean;
}

export class EmployeeRelationships {
    
@Expose()
    @Type(() => EmployeeRelationshipsProvider)
    @ValidateNested()
    provider: EmployeeRelationshipsProvider;
}

export class EmployeeRelationshipsProvider {
    
@Expose()
    @Type(() => EmployeeRelationshipsProviderData)
    @ValidateNested()
    data: EmployeeRelationshipsProviderData;
}

export class EmployeeRelationshipsProviderData {
    
@Expose()
    @IsString()
    type: string;
    
@Expose()
    @IsString()
    id: string;
}
"
`;

exports[`serialization primitive schemas + optionality 1`] = `
"import { Expose } from "class-transformer";
import { IsBoolean, IsInt, IsString } from "class-validator";

export class Response {
    
@Expose()
    @IsBoolean()
    optionalBool?: boolean;
    
@Expose()
    @IsBoolean()
    requiredBool: boolean;
    
@Expose()
    @IsInt()
    numeric?: number;
    
@Expose()
    @IsInt()
    requiredNum: number;
    
@Expose()
    @IsString()
    textual?: string;
    
@Expose()
    @IsString()
    requiredText: string;
}
"
`;

exports[`serialization proper name and ouput path 1`] = `
"export class SomeResponse2 {
}
"
`;

exports[`serialization string limits 1`] = `
"import { Expose } from "class-transformer";
import { MinLength, MaxLength, IsString } from "class-validator";

export class Response2 {
    
@Expose()
    @MinLength(1)
    @IsString()
    minLen?: string;
    
@Expose()
    @MaxLength(10)
    @IsString()
    maxLen?: string;
}
"
`;
